icu4c/source/i18n/messageformat2_function_registry.cpp

```diff
@@ -905,6 +906,149 @@ Formatter* StandardFunctions::DateTimeFactory::createFormatter(const Locale& loc
     return result;
 }
 
+static UDate tryPattern(const char* pat, const UnicodeString& sourceStr, UErrorCode& errorCode) {
+    LocalPointer<DateFormat> dateParser(new SimpleDateFormat(UnicodeString(pat), errorCode));
+    if (U_FAILURE(errorCode)) {
+        return 0;
+    }
+    return dateParser->parse(sourceStr, errorCode);
+}
+
+// From https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.md#date-and-time-operands :
+// "A date/time literal value is a non-empty string consisting of an ISO 8601 date, or
+// an ISO 8601 datetime optionally followed by a timezone offset."
+static UDate tryPatterns(const UnicodeString& sourceStr, UErrorCode& errorCode) {
+    if (U_FAILURE(errorCode)) {
+        return 0;
+    }
+    // Handle ISO 8601 datetime (tryTimeZonePatterns() handles the case
+    // where a timezone offset follows)
+    if (sourceStr.length() > 10) {
+        return tryPattern("YYYY-MM-dd'T'HH:mm:ss", sourceStr, errorCode);
+    }
+    // Handle ISO 8601 date
+    return tryPattern("YYYY-MM-dd", sourceStr, errorCode);
+}
+
+// See comment on tryPatterns() for spec reference
+static UDate tryTimeZonePatterns(const UnicodeString& sourceStr, UErrorCode& errorCode) {
+    if (U_FAILURE(errorCode)) {
+        return 0;
+    }
+    // 'zzzz' to handle the case where the timezone offset follows the ISO 8601 datetime
+    if (sourceStr.length() > 25) {
```

Comment:
Could you explain where the 25 came from? How do you conclude it need to > 25? and 25 is enough?

https://unicode.org/reports/tr35/tr35-dates.html#Contents show

zzzz | Pacific Daylight Time | The long specific non-location format. Where that is unavailable, falls back to the long localized GMT format ("OOOO").
-- | -- | --

and

OOOO | GMT-08:00 | The long localized GMT format.
-- | -- | --

Reply:
I suppose it doesn't really need to check the length at all, just whether the last character is 'Z'. Changed that in [c63c6b5](https://github.com/unicode-org/icu/pull/3012/commits/c63c6b52832c7251b93636ff33ca54d8161b271c).